libp2p/protocols/pubsub/gossipsub/scoring.nim

```diff
@@ -244,10 +252,11 @@ proc updateScores*(g: GossipSub) = # avoid async
 
     trace "updated peer's score", peer, score = peer.score, n_topics, is_grafted
 
-    if g.parameters.disconnectBadPeers and stats.score < g.parameters.graylistThreshold and
-        peer.peerId notin g.parameters.directPeers:
-      debug "disconnecting bad score peer", peer, score = peer.score
-      asyncSpawn(g.disconnectPeer(peer))
+    if not g.tryDisconnectBadPeer(peer, stats.score, g.parameters.graylistThreshold):
+      if peer.totalTraffic > 0:
```

Comment:
So this is only tested during heartbeats?

Reply:
yes, but I think we can do it in `punishInvalidMessage`, right?

Reply:
Even better, maybe we can do it in method `rpcHandler*` implementations.